server/src/main/java/io/crate/replication/logical/repository/PublisherRestoreService.java

```diff
@@ -0,0 +1,234 @@
+            metadataSnapshot = store.getMetadata(indexCommitRef.getIndexCommit());
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        } finally {
+            store.decRef();
+        }
+
+        // Identifies the seq no to start the replication operations from
+        var fromSeqNo = RetentionLeaseActions.RETAIN_ALL;
+
+        // Passing nodeclient of the leader to acquire the retention lease on leader shard
+        var retentionLeaseHelper = new RetentionLeaseHelper(request.subscriberClusterName(), nodeClient);
+        // Adds the retention lease for fromSeqNo for the next stage of the replication.
+        retentionLeaseHelper.addRetentionLease(
+            request.publisherShardId(),
+            fromSeqNo,
+            request.subscriberShardId(),
+            LogicalReplicationRepository.REMOTE_CLUSTER_REPO_REQ_TIMEOUT_IN_MILLI_SEC
+        );
+
+        /**
+         * At this point, it should be safe to release retention lock as the retention lease
+         * is acquired from the local checkpoint and the rest of the follower replay actions
+         * can be performed using this retention lease.
+         */
+        try {
+            retentionLock.close();
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+
+        var restoreContext = new RestoreContext(indexCommitRef, metadataSnapshot);
+        onGoingRestores.put(restoreUUID, restoreContext);
+
+        closableResources.add(restoreContext);
+        return restoreContext;
+    }
+
+    public void removeRestoreContext(String restoreUUID) {
+        var restoreContext = onGoingRestores.remove(restoreUUID);
+        if (restoreContext != null) {
+            restoreContext.close();
+        }
+    }
+
+
+    public static class RestoreContext implements Closeable {
+
+        private static final int INITIAL_FILE_CACHE_CAPACITY = 20;
+
+        private final Engine.IndexCommitRef indexCommitRef;
+        private final Store.MetadataSnapshot metadataSnapshot;
+        private final LinkedHashMap<String, IndexInput> currentFiles = new LinkedHashMap<>(INITIAL_FILE_CACHE_CAPACITY);
+
+        public RestoreContext(Engine.IndexCommitRef indexCommitRef,
+                              Store.MetadataSnapshot metadataSnapshot) {
+            this.indexCommitRef = indexCommitRef;
+            this.metadataSnapshot = metadataSnapshot;
+        }
+
+        public IndexInput openInput(Store store, String fileName) {
+            var currentIndexInput = currentFiles.getOrDefault(fileName, null);
+            if (currentIndexInput != null) {
+                return currentIndexInput.clone();
+            }
+
+            store.incRef();
+            try {
+                currentIndexInput = store.directory().openInput(fileName, IOContext.READONCE);
+            } catch (IOException e) {
+                throw new UncheckedIOException(e);
+            } finally {
+                store.decRef();
```

Comment:
I think we can safely remove the store inc/dec as this is already done for the resource by the caller at https://github.com/crate/crate/pull/11830/files#diff-0dc88d4d44588d4bd8c49bc188f04cb8dae6236c4061d21b28004709d5877238R100.
Additionally I think we should:
 - use the directory of the already defined `indexCommit` instead of the `store`
 - remove the `.clone()` for same reasons like removing the store ref counter, the resource will be consumed immediately by the caller.